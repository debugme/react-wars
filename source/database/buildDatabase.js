import mongoose from 'mongoose'

const buildDatabase = (options) => {
  mongoose.Promise = global.Promise
  mongoose.connect(options.mongodbUri)
  return mongoose
}

export default buildDatabase