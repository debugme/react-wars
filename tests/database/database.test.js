import _ from 'lodash'
import path from 'path'
import dotenv from 'dotenv'

import Character from '../../source/database/models/Character'
import buildDatabase from '../../source/database/buildDatabase'
import fields from './database.test.json'

describe('database operation handling', () => {

  let testDatabase = null

  beforeAll(() => {
    dotenv.config({ path: path.resolve(__dirname, '..', '.env') })
    testDatabase = buildDatabase(process.env)
  })

  afterAll((done) => {
    Character.collection.remove()
      .then(() => testDatabase.disconnect(done))

  })

  describe('create new document in characters collection', () => {
    it('should allow creation of a character from correct data', (done) => {
      const { name } = fields
      const keys = Object.keys(fields)
      const find = () => Character.findOne({ name })
      const check = (data) => expect(_.pick(data, keys)).toMatchSnapshot()
      new Character(fields).save()
        .then(find)
        .then(check)
        .then(done)
    })

    it('should disallow creation of a character from incorrect data', (done) => {
      const fields = { fruit: 'banana' }
      const check = (error) => expect(JSON.stringify(error)).toMatchSnapshot()
      new Character(fields).save()
        .catch(check)
        .then(done)
    })

  })

  describe('read document from characters collection', () => {

    it('should read the correct data for an existing character', (done) => {
      const { name } = fields
      const keys = Object.keys(fields)
      const check = (data) => expect(_.pick(data, keys)).toMatchSnapshot()
      Character.findOne({ name })
        .then(check)
        .then(done)
    })

    it('should return null for a non-existent character', (done) => {
      const name = 'Boba Fett'
      const check = (data) => expect(data).toMatchSnapshot()
      Character.findOne({ name })
        .then(check)
        .then(done)
    })
  })

  describe('update existing document in characters collection', () => {

    it('should update a character if correct values are provided', (done) => {
      const match = { name: fields.name }
      const update = { hair_color: 'black' }
      const options = { new: true }
      const keys = Object.keys(fields)
      const check = (data) => expect(_.pick(data, keys)).toMatchSnapshot()
      Character.findOneAndUpdate(match, update, options)
        .then(check)
        .then(done)
    })


  })

  describe('delete document from characters collection', () => {

    it('should allow an existing character to be deleted', (done) => {
      const { name } = fields
      const find = () => Character.findOne({ name })
      const check = (data) => expect(data).toMatchSnapshot()
      Character.findOneAndRemove({ name })
        .then(find)
        .then(check)
        .then(done)
    })

    it('should return null on delete of non-existent character', (done) => {
      const name = 'banana'
      const check = (data) => expect(data).toMatchSnapshot()
      Character.findOneAndRemove({ name })
        .then(check)
        .then(done)
    })

  })
})