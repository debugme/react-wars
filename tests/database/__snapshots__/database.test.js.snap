// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`database operation handling create new document in characters collection should allow creation of a character from correct data 1`] = `
Object {
  "birth_year": "29BBY",
  "eye_color": "brown",
  "hair_color": "brown",
  "height": 180,
  "is_male": true,
  "mass": 80,
  "name": "Han Solo",
  "skin_color": "fair",
}
`;

exports[`database operation handling create new document in characters collection should disallow creation of a character from incorrect data 1`] = `"{\\"errors\\":{\\"is_male\\":{\\"message\\":\\"Path \`is_male\` is required.\\",\\"name\\":\\"ValidatorError\\",\\"properties\\":{\\"type\\":\\"required\\",\\"message\\":\\"Path \`{PATH}\` is required.\\",\\"path\\":\\"is_male\\"},\\"kind\\":\\"required\\",\\"path\\":\\"is_male\\"},\\"name\\":{\\"message\\":\\"Path \`name\` is required.\\",\\"name\\":\\"ValidatorError\\",\\"properties\\":{\\"type\\":\\"required\\",\\"message\\":\\"Path \`{PATH}\` is required.\\",\\"path\\":\\"name\\"},\\"kind\\":\\"required\\",\\"path\\":\\"name\\"}},\\"message\\":\\"Characters validation failed\\",\\"name\\":\\"ValidationError\\"}"`;

exports[`database operation handling delete document from characters collection should allow an existing character to be deleted 1`] = `null`;

exports[`database operation handling delete document from characters collection should return null on delete of non-existent character 1`] = `null`;

exports[`database operation handling read document from characters collection should read the correct data for an existing character 1`] = `
Object {
  "birth_year": "29BBY",
  "eye_color": "brown",
  "hair_color": "brown",
  "height": 180,
  "is_male": true,
  "mass": 80,
  "name": "Han Solo",
  "skin_color": "fair",
}
`;

exports[`database operation handling read document from characters collection should return null for a non-existent character 1`] = `null`;

exports[`database operation handling update existing document in characters collection should update a character if correct values are provided 1`] = `Object {}`;
